src/com/amazon/corretto/crypto/provider/EvpHmac.java

```diff
@@ -0,0 +1,385 @@
+                assertInitialized();
+                synchronizedUpdateCtxArray(state.context, null, DO_NOT_INIT, src, offset, length);
+            })
+            .withDoFinal((ignored) -> {
+                assertInitialized();
+                final byte[] result = new byte[state.digestLength];
+                synchronizedDoFinal(state.context, result);
+                return result;
+            })
+            .withSinglePass((src, offset, length) -> {
+                assertInitialized();
+                final byte[] result = new byte[state.digestLength];
+                byte[] rawKey = state.encoded_key;
+                long evpMd = DO_NOT_REKEY;
+                if (state.needsRekey) {
+                    evpMd = state.evpMd;
+                }
+                synchronizedFastHmac(state.context, rawKey, evpMd, src, offset, length, result);
+                state.needsRekey = false;
+                return result;
+            });
+    }
+
+    @Override
+    protected int engineGetMacLength() {
+        return state.digestLength;
+    }
+
+    @Override
+    protected void engineInit(Key key, AlgorithmParameterSpec params)
+            throws InvalidKeyException, InvalidAlgorithmParameterException {
+        if (params != null) {
+            throw new InvalidAlgorithmParameterException("Params must be null");
+        }
+        if (!(key instanceof SecretKey)) {
+            throw new InvalidKeyException("Hmac uses expects a SecretKey");
+        }
+        state.setKey((SecretKey) key);
+        engineReset();;
+    }
+
+    @Override
+    protected void engineUpdate(byte input) {
+        buffer.update(input);
+    }
+
+    @Override
+    protected void engineUpdate(byte[] input, int offset, int len) {
+        buffer.update(input, offset, len);
+    }
+
+    @Override
+    protected byte[] engineDoFinal() {
+        return buffer.doFinal();
+    }
+
+    @Override
+    protected void engineReset() {
+        buffer.reset();
+    }
+
+    private void assertInitialized() {
+        if (state.key == null) {
+            throw new IllegalStateException("Mac not initialized");
+        }
+    }
+
+    @Override
+    public EvpHmac clone() throws CloneNotSupportedException {
+        EvpHmac cloned = (EvpHmac) super.clone();
+        cloned.state = cloned.state.clone();
+        cloned.buffer = cloned.buffer.clone();
+        cloned.configureLambdas();
+        return cloned;
+    }
+
+    private static final class HmacState implements Cloneable {
+        private SecretKey key;
+        private final long evpMd;
+        private final int digestLength;
+        private byte[] context = new byte[CONTEXT_SIZE];
+        private byte[] encoded_key;
+        boolean needsRekey = true;
+
+        private HmacState(long evpMd, int digestLength) {
+            this.evpMd = evpMd;
+            this.digestLength = digestLength;
+        }
+
+        private void setKey(SecretKey key) throws InvalidKeyException {
```

Comment:
it looks like the only caller of this is `engineInit`, but should we consider wrapping that call in a `synchronized` block to avoid races in case callers of `engineInit` invoke from multiple threads?

Reply:
`Mac` is distinctly not threadsafe in any form. The only reason we have synchronization at all is to protect the native layer from possible memory corruption and crashing the JVM. Misusing `Mac` over multiple threads will give incorrect answers no matter what you do, so it isn't worth the cost of synchronization to slightly harden it.